### PR TITLE
Refactor `std::array` usage in row group index writing in ORC

### LINKED ISSUE #6401: [BUG] write_index_stream in ORC writer indexing error
According to the comment https://github.com/rapidsai/cudf/pull/6342#issuecomment-700835023, there is an error in line
https://github.com/rapidsai/cudf/blob/86804ef3e0959af80d1d747fa00399164003342a/cpp/src/io/orc/writer_impl.cu#L868

In addition, there is a lack of test coverage for both this issue and https://github.com/rapidsai/cudf/issues/6333.
The test(s) should be added as part of this fix.

Note: the comment above also suggests replacing std::arrays in the `write_index_stream` with structs to improve code clarity. This can also be addressed in the same PR.

### PR DESCRIPTION
**Pull request by rgsl888prabhu**
Replaced std::array with named struct.
closes #6401

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: cpp/src/io/orc/writer_impl.cu [hunk 1]
```diff
@@ -38,6 +38,13 @@ namespace orc {
 using namespace cudf::io::orc;
 using namespace cudf::io;
 
+struct row_group_index_info {
+  int32_t pos;        // Position
+  int32_t blk_pos;    // Block Position
+  int32_t comp_pos;   // Compressed Position
+  int32_t comp_size;  // Compressed size
```
Comment (vuule) [inline cpp/src/io/orc/writer_impl.cu:45]: Feels like these should be uin32_t + a bool to indicate if the object is valid

File: python/cudf/cudf/tests/test_orc.py [hunk 2]
```diff
@@ -477,3 +477,16 @@ def test_int_overflow(tmpdir):
     df.to_orc(file_path)
 
     assert_eq(cudf.read_orc(file_path), df)
+
+
+def test_row_group_indexing():
```
Comment (vuule) [inline python/cudf/cudf/tests/test_orc.py:482]: This tests passes locally even without your changes :thinking:

**Review by vuule — CHANGES_REQUESTED** [review]

File: cpp/src/io/orc/writer_impl.cu (hunk 1 above)
Comment (rgsl888prabhu) [reply cpp/src/io/orc/writer_impl.cu:45]: If it is uint32_t, we will not be able to store have -1 as default value

Comment (rgsl888prabhu) [reply cpp/src/io/orc/writer_impl.cu:45]: And there are chances were one of the values are set and other is not, in that case it would be hard to discern.

Comment (vuule) [reply cpp/src/io/orc/writer_impl.cu:45]: yes, but then we don't need the default value, we can use a separate bool to indicate validity (unless I'm missing the meaning of -1).

File: python/cudf/cudf/tests/test_orc.py (hunk 2 above)
Comment (rgsl888prabhu) [reply python/cudf/cudf/tests/test_orc.py:482]: Yes it passes locally without my change as well. I think the test is meant to safeguard what was fixed in #6342

File: cpp/src/io/orc/writer_impl.cu (hunk 1 above)
Comment (vuule) [reply cpp/src/io/orc/writer_impl.cu:45]: my bad, looks like a record has two "layers" of validity. Darn.

File: python/cudf/cudf/tests/test_orc.py (hunk 2 above)
Comment (vuule) [reply python/cudf/cudf/tests/test_orc.py:482]: Can you please verify that?

Comment (rgsl888prabhu) [reply python/cudf/cudf/tests/test_orc.py:482]: I checked, with and without the change, it doesn't seems to fail even for very large number of rows, will try to get a failing test case.

Comment (rgsl888prabhu) [reply python/cudf/cudf/tests/test_orc.py]: couldn't find a test case where it would fail without fix from PR #6342 , so removing this test.

**Review by vuule — APPROVED** [review]

File: CHANGELOG.md [hunk 3]
```diff
@@ -88,6 +88,7 @@
 - PR #6786 Add nested type support to ColumnVector#getDeviceMemorySize
 - PR #6780 Move `cudf::cast` tests to separate test file
 - PR #6770 Support building decimal columns with Table.TestBuilder
+- PR #6807 Refactor std::array usage and add test coverage for row group index writing in ORC
```
Comment (vuule) [inline CHANGELOG.md:91]:
```suggestion
- PR #6807 Refactor `std::array` usage in row group index writing in ORC
```

**Review by karthikeyann — APPROVED** [review]

**Comment by vuule** [comment]
@kkraus14 This PR used to include Python changes so it requires a Python review. Can you please approve (as there are no Python changes)?

**Review by kkraus14 — APPROVED** [review]
Approving for Python anyway

**Comment by harrism** [comment]
Setting Okay to automerge.

**Comment by rgsl888prabhu** [comment]
@harrism should I remove change log ?

**Pull request merged**